exercises/handling-errors/README.md

```diff
@@ -0,0 +1,63 @@
+- Use non-retryable errors to fail an Activity
+- Locate the details of a failure in Temporal Workflows and Activities in the Event History
+
+Make your changes to the code in the `practice` subdirectory (look for `TODO` comments that will guide you to where you should make changes to the code). If you need a hint or want to verify your changes, look at the complete version in the `solution` subdirectory.
+
+## Setup
+
+You'll need two terminal windows for this exercise.
+
+1. In all terminals, change to the `exercises/handling-errors/practice` directory.
+2. In one terminal, run `npm install` to install dependencies.
+
+## Part A: Defining Errors for Activities
+
+In this part of the exercise, you will throw an Application Failure that will fail your Activities.
+
+Application Failures are used to communicate application-specific failures in Workflows and Activities. In Workflows, if you throw an `ApplicationFailure`, the Workflow Execution will fail. In Activities, you can either throw an `ApplicationFailure` or another Error to fail the Activity Task. 
+
+1. Edit the `activities.ts` file.
+2. Import `ApplicationFailure` from `@temporalio/common`.
+3. In the `sendBill` Activity, notice how we throw an Application Failure error if the charge is negative. The Application Failure includes a message as well as a details array with a list of details pertaining to the failure. Since we want our custom error data to be serialized and transmitted over the network, we must set our custom data in the `details` field. 
+4. Go to the `validateCreditCard` Activity. We want to throw an error if the entered credit card number does not have 16 digits. Replace the logged error with your own `ApplicationFailure`, following the pattern you saw in step 3.
+
+## Part B: Throw the Activity Failures in Your Workflow
+
+In this part of the exercise, you will throw your Activity Failures that will fail your Workflow Executions.
+
+When there is an error in an Activity, the failure is wrapped in an `ActivityFailure` and propagated back to the Workflow. This allows the Workflow to be aware of the failure and handle it appropriately. In the Workflow, this failure can be caught and handled using standard try-catch mechanisms.
+
+1. Edit the `workflows.ts` file.
+2. Add `ApplicationFailure` and `ActivityFailure` in your imports from `@temporalio/workflow` at the top of your file.
+3. Look at the call to the `validateCreditCard` Activity. We want to catch if there is an error in this Activity, so we check if the error is an instance of an `ActivityFailure` and the error's cause is an instance of an `ApplicationFailure`. If so, log the message of the error cause like `err.cause.message`. 
+4. Right now, if we have an Activity that has an invalid credit card number, the Activity will fail, but not the Workflow Execution. However, if we would like the failed Activity to fail the Workflow Execution, we need to make the Activity nonRetryable.
```

Comment:
That's wrong. `nonRetryable` on an `ApplicationFailure` thrown from an Activity determines if that activity should be retried or not. The Workflow will only "see" the error once the Activity can no longer be retried (either because `nonRetryable` was true, or that error's type is listed in the retry policy's `nonRetryableErrorTypes`, or because the timeout/max attempt set in the retry policy was exceeded).

Also, the failure of an Activity never directly causes a workflow failure. An `ActivityFailure` will cause failure of the Workflow if and only if the Workflow let that error bubble out of the workflow function (i.e. it doesn't catch the error, or catches and rethrows it).

In the exercise code, providing an incorrect credit card number will _not_ fail the workflow. You need to rethrow the error after logging the error message.

Reply:
Thank you @mjameswh . Can you please give us an example of how you would reword this? Is the behavior working on your end, and you don't like how it is framed? We tested it on our end, and it is behaving as expected. Thank you! :)

Reply:
CC @axfelix and @MasonEgger for visibility

Reply:
The way you fixed this in your most recent commit is correct.